### PR TITLE
run: Add trace sni

### PR DESCRIPTION
**Pull request by claudiamarcubina**
# Add trace_sni as image based gadget

This PR adds trace_sni as image based gadget.

### CODE REVIEW

File: gadgets/Makefile [hunk 1]
```diff
@@ -8,6 +8,7 @@ GADGETS = \
 	trace_dns \
 	trace_exec \
 	trace_mount \
+	trace_sni \
```
Comment (mauriciovasquezbernal) [inline gadgets/Makefile:11]: nit: please keep them ordered in the list.

File: gadgets/trace_sni/program.bpf.c [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+#include <linux/bpf.h>
+#include <linux/if_ether.h>
+#include <linux/ip.h>
+#include <linux/in.h>
+#include <linux/tcp.h>
+
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_endian.h>
+
+#include <gadget/macros.h>
```
[the remark this answers is not in the document]
Comment (claudiamarcubina) [reply gadgets/trace_sni/program.bpf.c:14]: Hi, Francis :) Thanks for the feedback!
Could you please have a look on how I've used buffer.h?

File: gadgets/trace_sni/program.bpf.c [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+	__u32 tid;
+	__u32 uid;
+	__u32 gid;
+	__u8 task[TASK_COMM_LEN];
+	__u8 name[TLS_MAX_SERVER_NAME_LEN];
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+} events SEC(".maps");
```
Comment (mauriciovasquezbernal) [inline gadgets/trace_sni/program.bpf.c:75]: This isn't needed anymore after using `GADGET_TRACER_MAP`. You should also use `gadget_reserve_buf` and `gadget_submit_buf` as done in https://github.com/inspektor-gadget/inspektor-gadget/pull/2323.

**Review by mauriciovasquezbernal — COMMENTED** [review]
It looks good. Compilation is failing but it seems related to buffer.h, I'll comment on https://github.com/inspektor-gadget/inspektor-gadget/pull/2323.

File: gadgets/trace_sni/gadget.yaml [hunk 4]
```diff
@@ -0,0 +1,56 @@
+name: sni
```
[the remark this answers is not in the document]
Comment (claudiamarcubina) [reply gadgets/trace_sni/gadget.yaml:1]: I think it's better to squash the commits indeed. Squashed.

File: gadgets/Makefile (hunk 1 above)
Comment (claudiamarcubina) [reply gadgets/Makefile]: Hi, Mauricio! Thanks for checking :) Done ✔️

File: gadgets/trace_sni/program.bpf.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (claudiamarcubina) [reply gadgets/trace_sni/program.bpf.c:14]: I think I got it. Could you please check?

**Comment by claudiamarcubina** [comment]
> struct {
> 	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
> } events SEC(".maps");

No problem. Do you mean to completely drop using buffer.h?

**Comment by claudiamarcubina** [comment]
> I wanted to test it and just get the following compilation errors:
> 
> ```
> $ sudo -E ./ig image build -t snis -f gadgets/trace_sni/gadget.yaml gadgets/trace_sni 
> INFO[0000] Experimental features enabled                
> Builder container logs start:
> clang -target bpf -Wall -g -O2  -D __TARGET_ARCH_x86 \
>         -c /work/program.bpf.c -I /usr/include/gadget/amd64/ -o /out/amd64.bpf.o
> clang -target bpf -Wall -g -O2  -D __TARGET_ARCH_arm64 \
>         -c /work/program.bpf.c -I /usr/include/gadget/arm64/ -o /out/arm64.bpf.o
> # No wasm file found. Nothing to do.
> In file included from /work/program.bpf.c:15:
> In file included from /usr/include/gadget/buffer.h:7:
> /usr/include/gadget/amd64/vmlinux.h:55:26: error: typedef redefinition with different types ('__kernel_ulong_t' (aka 'unsigned long') vs 'unsigned int')
> typedef __kernel_ulong_t __kernel_size_t;
> ```
> 
> There is a conflict between `buffer.h` and `socket-maps.h` when the gadget is a network one... For now, let's use perf buffer. Sorry for the back and forth, but I just discovered it recently ([#2323 (comment)](https://github.com/inspektor-gadget/inspektor-gadget/pull/2323#issuecomment-1867783854)), further investigation are needed to find a solution to use `buffer.h` with networking gadget.


No problem. Do you mean to completely drop using buffer.h?

**Comment by claudiamarcubina** [comment]
> I tested it and it works perfectly:
> 
> ```shell
> $ sudo -E ./ig image build -t snis -f gadgets/trace_sni/gadget.yaml gadgets/trace_sni 
> [sudo] Mot de passe de francis : 
> INFO[0000] Experimental features enabled                
> Successfully built docker.io/library/snis:latest@sha256:e828874fb8ca6bc21ae5ca31d0280c4e83267d3c162c0a725a2ab5ab7a888ac4
> $ sudo -E ./ig run snis         (remotes/origin/claudia/trace-sni-gadget-conversion) %
> INFO[0000] Experimental features enabled                
> RUNTIME.CONTAINERNAME                           PID                       TASK             NAME                                          
> test-trace-sni                                  16166                     wget             example.com
> ```
> 
> Thank you for dealing with the comments regarding `buffer.h`!

Thank you as well for your guidance! Merging now.

**Pull request merged**